pkg/cloudprovider/gcp/object_store.go

```diff
@@ -67,78 +67,88 @@ func (o *objectStore) Init(config map[string]string) error {
 		return errors.Errorf("credentials file pointed to by %s does not contain a private key", credentialsEnvVar)
 	}
 
-	client, err := google.DefaultClient(oauth2.NoContext, storage.DevstorageReadWriteScope)
-	if err != nil {
-		return errors.WithStack(err)
-	}
+	o.googleAccessID = jwtConfig.Email
+	o.privateKey = jwtConfig.PrivateKey
 
-	gcs, err := storage.New(client)
+	client, err := storage.NewClient(context.Background(), option.WithScopes(storage.ScopeReadWrite))
 	if err != nil {
 		return errors.WithStack(err)
 	}
-
-	o.gcs = gcs
-	o.googleAccessID = jwtConfig.Email
-	o.privateKey = jwtConfig.PrivateKey
+	o.client = client
 
 	return nil
 }
 
 func (o *objectStore) PutObject(bucket string, key string, body io.Reader) error {
-	obj := &storage.Object{
-		Name: key,
-	}
+	w := o.client.Bucket(bucket).Object(key).NewWriter(context.Background())
+	defer w.Close()
 
-	_, err := o.gcs.Objects.Insert(bucket, obj).Media(body).Do()
+	_, err := io.Copy(w, body)
 
 	return errors.WithStack(err)
 }
 
 func (o *objectStore) GetObject(bucket string, key string) (io.ReadCloser, error) {
-	res, err := o.gcs.Objects.Get(bucket, key).Download()
+	r, err := o.client.Bucket(bucket).Object(key).NewReader(context.Background())
 	if err != nil {
 		return nil, errors.WithStack(err)
 	}
 
-	return res.Body, nil
+	return r, nil
 }
 
 func (o *objectStore) ListCommonPrefixes(bucket string, delimiter string) ([]string, error) {
-	res, err := o.gcs.Objects.List(bucket).Delimiter(delimiter).Do()
-	if err != nil {
-		return nil, errors.WithStack(err)
+	q := &storage.Query{
+		Delimiter: delimiter,
 	}
 
-	// GCP returns prefixes inclusive of the last delimiter. We need to strip
-	// it.
-	ret := make([]string, 0, len(res.Prefixes))
-	for _, prefix := range res.Prefixes {
-		ret = append(ret, prefix[0:strings.LastIndex(prefix, delimiter)])
-	}
+	var res []string
+
+	iter := o.client.Bucket(bucket).Objects(context.Background(), q)
+
+	for {
+		obj, err := iter.Next()
+		if err == iterator.Done {
+			return res, nil
+		}
+		if err != nil {
+			return nil, errors.WithStack(err)
+		}
 
-	return ret, nil
+		if obj.Prefix != "" {
+			res = append(res, obj.Prefix[0:strings.LastIndex(obj.Prefix, delimiter)])
```

Comment:
I'm not sure we still need to take the substring? According to https://godoc.org/cloud.google.com/go/storage#Query:
> // ... Objects whose names,
>    // aside from the prefix, contain delimiter will have their name,
>    // truncated after the delimiter, returned in prefixes.

Reply:
the `Prefix` values always contain a trailing `delimiter` (just re-confirmed this), so we need to remove it to get backup names. The language in that godoc comment is super-confusing, but I'm fairly certain that how this works is:

- if there are any objects in the bucket that don't contain `delimiter` in their name, they'll be returned, with the `Name` field set and no `Prefix`. (note that we don't have any of these in Ark because all of our objects are prefixed with `<backup-name>/`).
-for any objects that *do* contain `delimiter`, a result will be returned with the `Name` field empty and `Prefix` set to the substring of their name up to and including `delimiter`.
-results containing only `Prefix` are de-duped.

Reply:
Ah, truncated *after* the delimiter. OK.